mongo/integration/client_side_encryption_test.go

```diff
@@ -0,0 +1,235 @@
+// Copyright (C) MongoDB, Inc. 2021-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+// +build cse
+
+package integration
+
+import (
+	"context"
+	"testing"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/event"
+	"go.mongodb.org/mongo-driver/internal/testutil"
+	"go.mongodb.org/mongo-driver/internal/testutil/assert"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// createDataKeyAndEncrypt creates a data key with the alternate name @keyName.
+// Returns a ciphertext encrypted with the data key as test data.
+func createDataKeyAndEncrypt(mt *mtest.T, keyName string) primitive.Binary {
+	mt.Helper()
+
+	kvClientOpts := options.Client().
+		ApplyURI(mtest.ClusterURI()).
+		SetReadConcern(mtest.MajorityRc).
+		SetWriteConcern(mtest.MajorityWc)
+
+	testutil.AddTestServerAPIVersion(kvClientOpts)
+
+	kmsProvidersMap := map[string]map[string]interface{}{
+		"local": {"key": localMasterKey},
+	}
+
+	kvClient, err := mongo.Connect(mtest.Background, kvClientOpts)
+	defer kvClient.Disconnect(mtest.Background)
+	assert.Nil(mt, err, "Connect error: %v", err)
+
+	err = kvClient.Database("keyvault").Collection("datakeys").Drop(mtest.Background)
+	assert.Nil(mt, err, "Drop error: %v", err)
+
+	ceOpts := options.ClientEncryption().
+		SetKmsProviders(kmsProvidersMap).
+		SetKeyVaultNamespace("keyvault.datakeys")
+
+	ce, err := mongo.NewClientEncryption(kvClient, ceOpts)
+	assert.Nil(mt, err, "NewClientEncryption error: %v", err)
+
+	dkOpts := options.DataKey().SetKeyAltNames([]string{keyName})
+	_, err = ce.CreateDataKey(mtest.Background, "local", dkOpts)
+	assert.Nil(mt, err, "CreateDataKey error: %v", err)
+
+	t, value, err := bson.MarshalValue("test")
+	assert.Nil(mt, err, "MarshalValue error: %v", err)
```

Comment:
[opt] you could just use `bsoncore` and avoid the error checking/`MarshalValue` call.

Reply:
Oh nice, done!